src/riskmatrix/views/asset.py

```diff
@@ -138,7 +140,7 @@ def populate_obj(self, obj: Asset) -> None:  # type:ignore[override]
             if assessment.risk_id in new_risk_ids:
                 continue
 
-            if assessment.modified is None:
+            if inspect(assessment).persistent and assessment.modified is None:
```

Comment:
I don't really see how an assessment could ever not be persistent at this point, since the only non-persistent assessments should be the ones added in the first loop above, so those should be covered by the `continue` statement directly above, since they always be in `new_risk_ids`.

Did you actually encounter a bug or is this defensive programming?

If you did, I believe the actual source of the bug to be the use of `if assessment.risk_id in new_risk_ids` instead of `if assessment.risk.id in new_risk_ids`. It's possible that `risk_id` will not be set until the object has been flushed and refreshed. We could also potentially change the constructor of `RiskAssessment` to set `risk_id` in addition to `risk`, as long as `risk.id` is not `None`.

Reply:
So the error message below appears if one tries to update a record on the __Assets__ with no selected catalog. And yes, one question would be if that's allowed, my assumption went' the way of assuming an update in that regard should be possible if the creation of one like that is possible.

```sh
Traceback (most recent call last):
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/waitress/channel.py", line 428, in service
    task.service()
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/waitress/task.py", line 168, in service
    self.execute()
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/waitress/task.py", line 434, in execute
    app_iter = self.channel.server.application(environ, start_response)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/fanstatic/publisher.py", line 228, in __call__
    return self.app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/fanstatic/injector.py", line 66, in __call__
    response = request.get_response(self.app)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/webob/request.py", line 1313, in send
    status, headers, app_iter = self.call_application(
                                ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/webob/request.py", line 1278, in call_application
    app_iter = application(self.environ, start_response)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/router.py", line 270, in __call__
    response = self.execution_policy(environ, self)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid_retry/__init__.py", line 127, in retry_policy
    response = router.invoke_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/router.py", line 245, in invoke_request
    response = handle_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid_tm/__init__.py", line 178, in tm_tween
    raise exc from None
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid_tm/__init__.py", line 145, in tm_tween
    response = handler(request)
               ^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/tweens.py", line 43, in excview_tween
    response = _error_handler(request, exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/tweens.py", line 17, in _error_handler
    reraise(*exc_info)
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/util.py", line 733, in reraise
    raise value
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
               ^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/router.py", line 143, in handle_request
    response = _call_view(
               ^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/view.py", line 674, in _call_view
    response = view_callable(context, request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/config/views.py", line 151, in __call__
    return view(context, request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/config/views.py", line 170, in attr_view
    return view(context, request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/config/views.py", line 196, in predicate_wrapper
    return view(context, request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/viewderivers.py", line 319, in secured_view
    return view(context, request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/viewderivers.py", line 512, in csrf_view
    return view(context, request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/pyramid/viewderivers.py", line 427, in rendered_view
    result = view(context, request)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/christof/source/riskmatrix/src/riskmatrix/views/asset.py", line 311, in edit_asset_view
    form.populate_obj(asset)
  File "/home/christof/source/riskmatrix/src/riskmatrix/views/asset.py", line 142, in populate_obj
    session.delete(assessment)
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3498, in delete
    self._delete_impl(state, instance, head=True)
  File "/home/christof/source/riskmatrix/venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3505, in _delete_impl
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Instance '<RiskAssessment at 0x7f3ca401dc90>' is not persisted
```

Reply:
Maybe the culprit is `query = query.filter(Risk.catalog_id.in_(obj.catalog_ids))` then, this should result in a query that yields no risks at all, so no new risks are added, but if you get non-persistent assessments with no new risks that's clearly not working.

The problem with your workaround is that while it avoids the exception, the assessments will get persisted anyways, so you're adding risks that weren't meant to be added. There's `session.expunge` to remove non-persistent objects from the session, but you could also do a `session.flush` before the second loop. But either way the second loop should only ever be removing persistent instances that were part of `existing_risk_ids`, and are now being removed because the risk catalog they belong to has been de-selected.

The other problem I mentioned probably still plays into this as well.